### ISSUE TITLE
Knuth`s magical 3%

### ISSUE THREAD

**Issue by Bob-Murphy** [official-antistasi-community/A3-Antistasi#193]
Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%. 
[15:19] Spoffy: Huh. Interesting. I just tested a hunch of mine (that distanceSqr is much faster than distance). And... Not really. Mildly faster to the point of being negligible.
[15:19] Spoffy: TIL.
[15:22] Frost'sBite: Multiple that 0.001 by being promotional square by number of players and ai loaded
[15:39] Spoffy: Mmm, which is Knuth's 3%.
[15:39] Spoffy: Most of the time, it's not worth optimising.
[15:40] Spoffy: But when you're in a loops that's called once for every player, ai unit pair (worst case Antistasi, that's 50 * 300 or so, or 15,000)... That's no longer tiny.
[15:41] Spoffy: Now, let's say you really are having a bad day, and call that loop every frame.
[15:41] Spoffy: 0.001ms becomes 15ms of cost per frame.
[15:42] Spoffy: At 30FPS, you have 33ms to do everything in the game
[15:42] Spoffy: You've just spent half of your total game performance budget on a single loop.
[15:50] Spoffy: Also, this conversation has convinced me how we should profile Antistasi.
[15:50] Spoffy: We should add a command that measures how frequently a function or loop is called.
[15:51] Spoffy: It's not quite enough to tell us what's actually causing lag - for example, a function that does a lot of different things in one frame once is worse than a small function running once every frame.

TLDR:
Add a command that measures how frequently a function or loop is called

**Comment by wurzel0701** [comment]
There is the `diag_activeScripts` command, which shows the currently active scripts, but that may not fully show all needed info

**Comment by AlexTriada** [comment]
it’s not clear, but very interesting))) in my opinion, the mission needs to be put in order, because a lot of it is just rubbish, for example, UPSMON scripts, etc., all scripts need to be rewritten, which is not a problem for the team, and this is not so much about optimization, how much about dusting out and finding bugs and a bad code that leads to brakes

**Comment by Bob-Murphy** [comment]
Regarding for example UPSMON, we are thinking about maybe getting rid of it and eventually replace it with VCOM or something else. Until now we have not really evaluated all the options.
But yes, this 3% thing definitely is about finding said rubbish in the code, clear it up and hopefully enhance the performance and besides not understanding much, I really am exited for what we may find.

**Comment by Bob-Murphy** [comment]
@Spoffy 
When we change the AI system we at least should implement all the tracking of AI related stuff so we don't have to touch it again. Correct me if I'm wrong.

**Comment by Spoffy** [comment]
We should. We should also leave this issue open, I'd totally forgotten about the idea, but I still see the value of it.

**Comment by CalebSerafin** [comment]
UPSMON to VCOM now has its own issue thread. https://github.com/official-antistasi-community/A3-Antistasi/issues/448

Knuths 3% may not matter, but pre-optimising (_planning_) by having **_Good Design_** is the only way a game with SQF will run at 60fps. That does not usually save 3%, but at least 5-10x. Cutting down n milliseconds to n microseconds.
